### ISSUE TITLE
Can't post form-data, but ruby and nodejs works

### ISSUE THREAD

**Issue by birbird** [a85/POSTMan-Chrome-Extension#1273]
I dig this for a long time and found that it may not be my server problem.
I have used the postman inteceptor. And remove the restricted http headers and without postman interceptor, the result is the same.

The http code generated by postman:

```
POST /api/users/sign_up.json HTTP/1.1
Host: localhost:3000
Accept: application/json, text/javascript, */*; q=0.01
Accept-Encoding: gzip, deflate
Accept-Language: zh-CN,zh;q=0.8
Content-Type: multipart/form-data; boundary=----WebKitFormBoundary7MA4YWxkTrZu0gW
User-Agent: Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.134 Safari/537.36
X-Requested-With: XMLHttpRequest
Cache-Control: no-cache
Postman-Token: a93ff723-f199-e9bb-d2d1-fc1bef990f19

----WebKitFormBoundary7MA4YWxkTrZu0gW
Content-Disposition: form-data; name="user[mobile_number]"

13812345678
----WebKitFormBoundary7MA4YWxkTrZu0gW
Content-Disposition: form-data; name="user[password]"

abcd
----WebKitFormBoundary7MA4YWxkTrZu0gW
Content-Disposition: form-data; name="user[name]"

DD
----WebKitFormBoundary7MA4YWxkTrZu0gW
Content-Disposition: form-data; name="user[gender]"

male
----WebKitFormBoundary7MA4YWxkTrZu0gW
Content-Disposition: form-data; name="user[birthday]"

20001010
----WebKitFormBoundary7MA4YWxkTrZu0gW
Content-Disposition: form-data; name="user[picture]"; filename="4610b912c8fcc3ceb9f2b3009345d688d43f20d6.jpg"
Content-Type: image/jpeg


----WebKitFormBoundary7MA4YWxkTrZu0gW
Content-Disposition: form-data; name="verify_code"

790
----WebKitFormBoundary7MA4YWxkTrZu0gW
```

When post the data to my server, the server does not get the data and throws 

```
{"error":"user is missing, user[mobile_number] is missing, user[password] is missing, user[name] is missing, user[gender] is missing, user[birthday] is missing, user[picture] is missing, verify_code is missing"}
```

But if I run the ruby and nodejs code generated by postman, they all works well.
The ruby code:

```
require 'uri'
require 'net/http'

url = URI("http://localhost:3000/api/users/sign_up.json")

http = Net::HTTP.new(url.host, url.port)

request = Net::HTTP::Post.new(url)
request["accept"] = 'application/json, text/javascript, */*; q=0.01'
request["accept-encoding"] = 'gzip, deflate'
request["accept-language"] = 'zh-CN,zh;q=0.8'
request["content-type"] = 'multipart/form-data; boundary=---011000010111000001101001'
request["user-agent"] = 'Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.134 Safari/537.36'
request["x-requested-with"] = 'XMLHttpRequest'
request.body = "-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"user[mobile_number]\"\r\n\r\n13812345678\r\n-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"user[password]\"\r\n\r\nabcd\r\n-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"user[name]\"\r\n\r\nDD\r\n-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"user[gender]\"\r\n\r\nmale\r\n-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"user[birthday]\"\r\n\r\n20001010\r\n-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"user[picture]\"; filename=\"4610b912c8fcc3ceb9f2b3009345d688d43f20d6.jpg\"\r\nContent-Type: image/jpeg\r\n\r\n\r\n-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"verify_code\"\r\n\r\n790\r\n-----011000010111000001101001--"

response = http.request(request)
puts response.read_body
```

The nodejs code:

```
var http = require("http");

var options = {
  "method": "POST",
  "hostname": "localhost",
  "port": "3000",
  "path": "/api/users/sign_up.json",
  "headers": {
    "accept": "application/json, text/javascript, */*; q=0.01",
    "accept-encoding": "gzip, deflate",
    "accept-language": "zh-CN,zh;q=0.8",
    "content-type": "multipart/form-data; boundary=---011000010111000001101001",
    "user-agent": "Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.134 Safari/537.36",
    "x-requested-with": "XMLHttpRequest"
  }
};

var req = http.request(options, function (res) {
  var chunks = [];

  res.on("data", function (chunk) {
    chunks.push(chunk);
  });

  res.on("end", function () {
    var body = Buffer.concat(chunks);
    console.log(body.toString());
  });
});

req.write("-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"user[mobile_number]\"\r\n\r\n13812345678\r\n-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"user[password]\"\r\n\r\nabcd\r\n-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"user[name]\"\r\n\r\nDD\r\n-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"user[gender]\"\r\n\r\nmale\r\n-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"user[birthday]\"\r\n\r\n20001010\r\n-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"user[picture]\"; filename=\"4610b912c8fcc3ceb9f2b3009345d688d43f20d6.jpg\"\r\nContent-Type: image/jpeg\r\n\r\n\r\n-----011000010111000001101001\r\nContent-Disposition: form-data; name=\"verify_code\"\r\n\r\n790\r\n-----011000010111000001101001--");
req.end();
```

**Comment by abhijitkane** [comment]
@birbird Can you try choosing 'x-www-form-urlencoded' instead of form-data?

**Comment by birbird** [comment]
@abhijitkane Thanks for your quick reply.
I can use x-www-form-urlencoded if I remove the picture field, it works well.
But if I want to do a full test including all fields, I still have to use form-data.

**Comment by snowleung** [comment]
i can see two -- symbol at tail.
it's the reason?

**Comment by sebjamesfd** [comment]
This is due to the boundary within the form data not being generated correctly.

According to specifications (https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html), the boundary in the form-data must be prepended with two extra dashes '--', and for the final boundary must also be appended by two extra dashes '--'

**Comment by bfarayev** [comment]
Postman generated POST request using form-data and it works fine for me in Python + requests.

**Comment by sebjamesfd** [comment]
@bfarayev That's because python requests generates the boundary string, not Postman. Try generating a curl request.

**Comment by sdnts** [comment]
Guys, this should be fixed in new versions of Postman (4.0+), and you should be able to send form-data bodies.
